### PR TITLE
Simplify double precision input and module conversion

### LINKED ISSUE #13000: Expose hooks to process input args in precision plugins
### Motivation

#12983 attempted to fix an issue where the evaluation step methods in the DDPStrategy would call the `LightningModule.x_step `methods directly instead of going through the `LightningDoublePrecisionModule.x_step` methods.

This introduced a typing issue where we can't guarantee that `self.model` implements the `x_step` methods.

 
### Pitch

Treat #12983 as a temporary fix and refactor away the need to have a precision wrapper. 
All the precision wrapper does is [process the input arguments](https://github.com/PyTorchLightning/pytorch-lightning/blob/3d74c90e8802f91dcd6dc1492123eb5b87c49db9/pytorch_lightning/plugins/precision/double.py#L51-L55).  
This input processing could be done directly in the strategy. 

Before

```py
    def test_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
        with self.precision_plugin.test_step_context():
            return self.MAYBE_WRAPPER.test_step(*args, **kwargs)
```

After

```py
    def test_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
        with self.precision_plugin.test_step_context():
            return self.LIGHTNING_MODULE.test_step(self.precision_plugin.process_step_inputs(*args, **kwargs))
```

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @rohitgr7 @carmocca @akihironitta

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Refactors the double precision logic for converting inputs and model weight conversion. This needs to be standardized and simplified for supporting half-precision: #18193

This is only possible after #17531 
Fixes https://github.com/Lightning-AI/lightning/issues/17655
Closes #13000 (no longer needed)

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @justusschock @awaelchli @tchaton @carmocca

### CODE REVIEW

File: src/lightning/pytorch/plugins/precision/double.py [hunk 1]
```diff
@@ -12,80 +12,24 @@
...
 from torch import Tensor
-from torch.optim import Optimizer
 
-import lightning.pytorch as pl
 from lightning.fabric.plugins.precision.utils import _convert_fp_tensor
-from lightning.pytorch.overrides.base import _LightningPrecisionModuleWrapperBase
 from lightning.pytorch.plugins.precision.precision_plugin import PrecisionPlugin
 
 
-class LightningDoublePrecisionModule(_LightningPrecisionModuleWrapperBase):
```
Comment (awaelchli) [inline src/lightning/pytorch/plugins/precision/double.py:29]: need to deprecate this

File: src/lightning/pytorch/plugins/precision/double.py [hunk 2]
```diff
@@ -12,28 +12,76 @@
...
 import lightning.pytorch as pl
 from lightning.fabric.plugins.precision.utils import _convert_fp_tensor
-from lightning.pytorch.overrides.base import _LightningPrecisionModuleWrapperBase
+from lightning.fabric.utilities.device_dtype_mixin import _DeviceDtypeModuleMixin
 from lightning.pytorch.plugins.precision.precision_plugin import PrecisionPlugin
+from lightning.pytorch.utilities.rank_zero import rank_zero_deprecation
 
 
-class LightningDoublePrecisionModule(_LightningPrecisionModuleWrapperBase):
+class DoublePrecisionPlugin(PrecisionPlugin):
```
Comment (awaelchli) [inline src/lightning/pytorch/plugins/precision/double.py:29]: Reviewer: Sorry for the confusing diff. I moved the `LightningDoublePrecisionModule` to the bottom of the file, and this shows up as deleting and adding the precision plugin at the top.

File: src/lightning/pytorch/plugins/precision/double.py [hunk 3]
```diff
@@ -12,28 +12,76 @@
...
     """LightningModule wrapper which converts incoming floating point data in ``*_step`` and ``forward`` to double
     (``torch.float64``) precision.
 
+    Deprecated. Use :meth:`~lightning.pytorch.core.hooks.ModelHooks.configure_model` instead.
+
     Args:
         pl_module: the model to wrap
     """
 
+    def __init__(self, pl_module: "pl.LightningModule") -> None:
```
Comment (awaelchli) [inline src/lightning/pytorch/plugins/precision/double.py:73]: Previously, the class subclassed `_LightningPrecisionModuleWrapperBase`, but I removed it and inlined the remaining code here.

File: src/lightning/pytorch/plugins/precision/double.py [hunk 4]
```diff
@@ -12,28 +12,76 @@
...
+        super().__init__()
+        rank_zero_deprecation(
+            f"The `{type(self).__name__}` is deprecated and no longer needed. Convert the inputs to the `*_step`"
```
Comment (carmocca) [inline src/lightning/pytorch/plugins/precision/double.py:76]: Can you add a deprecation test?

File: src/lightning/pytorch/strategies/strategy.py [hunk 5]
```diff
@@ -152,6 +152,7 @@ def setup_precision_plugin(self) -> None:
         model, optimizers, lr_scheduler_configs = self.precision_plugin.connect(
             self.model, self.optimizers, self.lr_scheduler_configs
         )
+        model = self.precision_plugin.convert_module(model)
```
Comment (carmocca) [inline src/lightning/pytorch/strategies/strategy.py:155]: What about doing it in connect? This would resolve https://github.com/Lightning-AI/lightning/issues/17655

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/plugins/precision/double.py [hunk 6]
```diff
@@ -12,28 +12,76 @@
...
+
+    def convert_input(self, data: Any) -> Any:
+        return apply_to_collection(data, function=_convert_fp_tensor, dtype=Tensor, dst_type=torch.double)
+
+
+class LightningDoublePrecisionModule(_DeviceDtypeModuleMixin, nn.Module):
     """LightningModule wrapper which converts incoming floating point data in ``*_step`` and ``forward`` to double
     (``torch.float64``) precision.
 
+    Deprecated. Use :meth:`~lightning.pytorch.core.hooks.ModelHooks.configure_model` instead.
```
Comment (Borda) [inline src/lightning/pytorch/plugins/precision/double.py:67]:
```suggestion
    .. deprecated:: Use :meth:`~lightning.pytorch.core.hooks.ModelHooks.configure_model` instead.
```
see: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/strategies/strategy.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning/pytorch/strategies/strategy.py]: Yes that works too!

File: src/lightning/pytorch/plugins/precision/double.py (hunk 4 above)
Comment (awaelchli) [reply src/lightning/pytorch/plugins/precision/double.py:76]: There already is one :)

**Review by lantiga — APPROVED** [review]
LGTM!
